File: frontend/apps/main/src/types/manifesto.ts
export interface Manifesto {
  id: string
  family_id: string
  current_version_id: string | null
  status: 'draft' | 'active' | 'archived'
  signing_deadline: string | null
  created_by: string
  created_at: string
  current_version: ManifestoVersion | null
  signatures: ManifestoSignature[]
  rejections: ManifestoRejection[]
}

export interface ManifestoVersion {
  id: string
  version_number: number
  template_id: string
  title: string
  body: string
  change_type: 'initial' | 'minor' | 'major'
  trackable_clause_indices: number[] | null
  signed_at: string | null
  created_by: string
  created_at: string
}

export interface ManifestoSignature {
  id: string
  user_id: string
  signature_data: string | null // null = tap-to-consent
  signed_at: string
}

export interface ManifestoRejection {
  id: string
  user_id: string
  reason: string | null
  created_at: string
}

export interface ManifestoDashboardSummary {
  manifesto_id: string | null
  title: string
  total_members: number
  signed_count: number
  status: string
}

export interface ManifestoFeedback {
  id: string
  user_id: string
  content: string
  is_read: boolean
  created_at: string
}

export interface ManifestoVersionHistoryItem {
  id: string
  version_number: number
  change_type: string
  title: string
  created_by: string
  created_at: string
}

export interface UnsignedManifestoCheck {
  has_unsigned: boolean
  manifesto_id: string | null
  title: string | null
}

import type { Component } from 'vue'

export interface TemplateDefinition {
  id: string
  nameKey: string
  lang: 'zh' | 'en'
  type: 'classic' | 'modern'
  component: Component
}

// ---------------------------------------------------------------------------
// Flow visualization types
// ---------------------------------------------------------------------------

export type FlowState = 'draft' | 'signing' | 'rejected' | 'modifying' | 'effective' | 'expired'

export type MemberSigningStatus =
  | 'signed'
  | 'confirmed'
  | 'pending_sign'
  | 'pending_confirm'
  | 'rejected'
  | 'expired'

export interface FlowMemberState {
  userId: string
  displayName: string
  role: 'owner' | 'member' | 'child'
  status: MemberSigningStatus
  isCurrentUser: boolean
  signatureData: string | null | undefined
  rejectionReason: string | null | undefined
  signedAt: string | null
  rejectedAt: string | null
}

export interface FlowDeadlineInfo {
  expired: boolean
  remaining: string | null
}

export interface FlowSigningProgress {
  signed: number
  total: number
}

export interface FlowCreatedNodeData {
  changeType: 'initial' | 'minor' | 'major'
  versionNumber: number
  creatorName: string
  createdAt: string
  dimmed?: boolean
}

export interface FlowPendingNodeData {
  deadline: string | null
  deadlineExpired: boolean
  dimmed?: boolean
}

export interface FlowVersionUpdateNodeData {
  updaterName: string
  currentVersion: number
  nextVersion: number
  dimmed?: boolean
}

File: server/apps/backend/app/schemas/manifesto.py
"""Pydantic schemas for family manifesto feature."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from apps.backend.app.schemas.base import SnowflakeBase

# ---------------------------------------------------------------------------
# Request schemas
# ---------------------------------------------------------------------------


class ManifestoCreateRequest(BaseModel):
    template_id: str
    title: str = Field(max_length=200)
    body: str
    signing_deadline: datetime | None = None
    trackable_clause_indices: list[int] | None = None


class ManifestoPublishRequest(BaseModel):
    title: str | None = None
    body: str | None = None
    change_type: str = Field(pattern=r"^(minor|major)$")
    trackable_clause_indices: list[int] | None = None


class ManifestoSignRequest(BaseModel):
    signature_data: str | None = None


class ManifestoFeedbackCreateRequest(BaseModel):
    content: str


# ---------------------------------------------------------------------------
# Response schemas
# ---------------------------------------------------------------------------


class ManifestoSignatureItem(SnowflakeBase):
    id: int
    user_id: int
    signature_data: str | None
    signed_at: datetime


class ManifestoVersionItem(SnowflakeBase):
    id: int
    version_number: int
    template_id: str
    title: str
    body: str
    change_type: str
    trackable_clause_indices: list[int] | None
    signed_at: datetime | None
    created_by: int
    created_at: datetime


class ManifestoResponse(SnowflakeBase):
    id: int
    family_id: int
    current_version_id: int | None
    status: str
    signing_deadline: datetime | None
    created_by: int
    created_at: datetime
    current_version: ManifestoVersionItem | None = None
    signatures: list[ManifestoSignatureItem] = []


class ManifestoDashboardSummaryResponse(SnowflakeBase):
    manifesto_id: int
    title: str
    total_members: int
    signed_count: int
    status: str


class ManifestoFeedbackResponse(SnowflakeBase):
    id: int
    user_id: int
    content: str
    is_read: bool
    created_at: datetime


class ManifestoVersionHistoryItem(SnowflakeBase):
    id: int
    version_number: int
    change_type: str
    title: str
    created_by: int
    created_at: datetime


class UnsignedManifestoCheckResponse(SnowflakeBase):
    has_unsigned: bool
    manifesto_id: int | None = None
    title: str | None = None


class ChildManifestoResponse(SnowflakeBase):
    manifesto_id: int
    title: str
    body: str
    template_id: str
    signed: bool
    signer_names: list[str] = []


class ChildTrackableClausesResponse(BaseModel):
    has_trackable: bool
    trackable_clause_indices: list[int] = []
